torch_xla/csrc/runtime/xla_mlir_debuginfo_helper.cc

```diff
@@ -0,0 +1,92 @@
+#include "torch_xla/csrc/runtime/xla_mlir_debuginfo_helper.h"
+
+#include <cstring>
+#include <string>
+#include <utility>
+
+#include "absl/log/log.h"
+
+namespace torch_xla {
+namespace runtime {
+
+namespace {
+
+// Defined in torch_xla/experimental/xla_mlir_debuginfo.py
+static constexpr char XLA_MLIR_DEBUGINFO_BEGIN[] = "<XLA_MLIR_DEBUGINFO_BEGIN>";
+static constexpr char XLA_MLIR_DEBUGINFO_END[] = "<XLA_MLIR_DEBUGINFO_END>";
+
+class PrepareXlaMlirDebuginfoPass : public mlir::OperationPass<mlir::ModuleOp> {
+ public:
+  explicit PrepareXlaMlirDebuginfoPass()
+      : mlir::OperationPass<mlir::ModuleOp>::OperationPass(
+            mlir::TypeID::get<PrepareXlaMlirDebuginfoPass>()) {}
+
+  ~PrepareXlaMlirDebuginfoPass() override = default;
+
+  void runOnOperation() override {
+    mlir::MLIRContext* context = &getContext();
+    getOperation().walk([&](mlir::Operation* op) {
+      llvm::SmallVector<std::string> debuginfos;
+      ExtractXlaMlirDebuginfo(op->getLoc(), debuginfos);
+
+      if (!debuginfos.empty()) {
+        // If multiple debuginfos are found (which should be an exception),
+        // pick arbitrary one and discard the rest;
+        const std::string& debuginfo = debuginfos[0];
```

Comment:
`write_mlir_debuginfo` assumes the string data can overwrite all the other debuginfo associated with the tensor, including the torch_xla default one and the previous data from the same API call. So if the user/developer still wants the source line info with the debuginfo passed to this API, they will need to "encode" source line info explicitly in the data like `write_mlir_debuginfo(tensor, '{"debuginfo": ..., "source": ...}')`

Reply:
The entire encoding would be propagated to MLIR in the form like
```
stablehlo.add ... #loc("{\"debuginfo\": ..., \"source\": ...}")
```

Reply:
@chunnienc Thanks for the answers! To confirm my understanding, it is `_set_xla_custom_op_name_prefix`'s behavior to overwrite the existing debugging info.

Reply:
`_set_xla_custom_op_name_prefix` appends it to the existing torch_xla debug info (while only one prefix can be set at a time, so multiple calls would be overwritten). So the the original debug info for a tensor isthe op name e.g. `aten.add`, and after calling the `write_mlir_debuginfo(t, "HELLOWORLD")` the debug info would become `<XLA_MLIR_DEBUGINFO_BEGIN>HELLOWORLD<XLA_MLIR_DEBUGINFO_END>aten.add`

Reply:
The multiple debuginfos case happens when the stablehlo pass fuses ops in MLIR and their debug info into a mlir `FusedLoc`, that is why I'd say it is an exception because it is out of our control.

Reply:
Got it, thanks for the detailed explanation!